### ISSUE TITLE
Passing a string that contains a semi-colon doesn't properly get encoded

### ISSUE THREAD

**Issue by bobber205** [angular/angular.js#9224]
Discovered this today in a rails app as the backend (since it totally barfs on unescaped semis)

I had a string such as "test;test"

passed it to params in $http post() and put() and got 

```
test;test instead of 
```

```
%3B
```

Shouldn't it encode?

**Comment by schmod** [comment]
I work with a platform that has a (partial) implementation of [Matrix URIs](http://www.w3.org/DesignIssues/MatrixURIs.html).  

Admittedly, Matrix URI never became a standard, and is not in widespread use.  However, there are definitely existing implementations that rely on unescaped semicolons being in the `path`.

The `query` seems to be much more of a grey area, although the older HTML [specifications](http://www.w3.org/MarkUp/html-spec/html-spec_foot.html#FOOT26) [encouraged](http://www.w3.org/TR/REC-html40/appendix/notes.html#h-B.2.2) CGI developers to support semicolon-delimited query strings.

IMO, there's enough precedent to make me apprehensive about automatically escaping semicolons in _any_ URI component.  If a developer is interacting with an idiosyncratic server implementation, it seems to make the most sense to fix the server, or write an interceptor to "sanitize" URLs before they are sent to the server.  

**HOWEVER**

@bobber205's issue is a very specific and isolated use case:  Putting semicolons inside of a single query parameter value.  In this *exact* situation (and only this situation), I actually think that it is appropriate to escape semicolons, as server implementations might misinterpret an unescaped semicolon as a delimiter.

In summary:  
* `path`: Don't escape semicolons.  Breaks matrix parameters, and no evidence that the current behavior breaks anything.
* `query`: Escaping the entire query string would be unwise, as it breaks an established (albeit archaic) expectation for servers to be able to handle semicolon-delimited query strings.
* `query` values:  Probably safe to escape, especially if the server interprets semicolons as query delimiters.  Passing complex data types (ie. JSON) via query parameters is a bad idea, "real-world" string values often contain semicolons, and it seems like a good idea to escape them.

**Comment by bobber205** [comment]
Thanks for all the hardwork! <3
